### PR TITLE
Use new Json endpoint for chrome driver

### PR DESCRIPTION
**Pull request by Gurdarrag**

### CODE REVIEW

File: src/Driver/ChromeDriver/JsonDownloader.php
Comment (dbrekelmans) [inline src/Driver/ChromeDriver/JsonDownloader.php:1]: Is anything here different than in `ChromeDriver/Downloader` except for the `DOWNLOAD_ENDPOINT` constant?

If not, then I'd suggest creating a private function in the original download to determine the endpoint based on the driver version, and using that function where this constant is used at the moment

File: src/Driver/VersionResolver.php [hunk 1]
```diff
@@ -11,6 +11,8 @@
 
 interface VersionResolver
 {
+    public const CHROME_MAJOR_VERSION_ENDPOINT_BREAKPOINT = 115;
```
Comment (dbrekelmans) [inline src/Driver/VersionResolver.php:14]: This const does not belong in the generic driver interface, but in the chrome specific one

File: src/Command/Input/OperatingSystemOption.php [hunk 2]
```diff
@@ -82,6 +82,7 @@ public static function value(InputInterface $input)
             );
         }
 
+        /** @psalm-var string $value */
```
Comment (dbrekelmans) [inline src/Command/Input/OperatingSystemOption.php:85]: This should not be necessary due to the `is_string` check on line 71... are you getting a psalm error here?

File: psalm.xml [hunk 3]
```diff
@@ -1,6 +1,5 @@
 <?xml version="1.0"?>
 <psalm
-        totallyTyped="true"
```
Comment (dbrekelmans) [inline psalm.xml:3]: Why did you change this?

File: src/Driver/ChromeDriver/JsonVersionResolver.php [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+        $browserName = $browser->name();
+
+        return ($browserName->equals(BrowserName::GOOGLE_CHROME()) || $browserName->equals(BrowserName::CHROMIUM()))
+            && $browser->version()->major() >= VersionResolverInterface::CHROME_MAJOR_VERSION_ENDPOINT_BREAKPOINT;
+    }
+
+    /**
+     * In case of handling with Chrome from Dev or Canary channel we will then take beta ChromeDriver
+     */
+    private function getBrowserVersionEndpoint(Browser $browser): string
```
Comment (dbrekelmans) [inline src/Driver/ChromeDriver/JsonVersionResolver.php:89]: This function is not used anywhere, what's the intention here?

**Review by dbrekelmans — CHANGES_REQUESTED** [review]

File: src/Driver/ChromeDriver/JsonDownloader.php
Comment (Gurdarrag) [reply src/Driver/ChromeDriver/JsonDownloader.php:1]: Yeah the downloaded archive is not build the same as previously so the tmp path is not the same
But download / downloadArchive and getFilename are the same.
Maybe an Abstract class on the top or JsonDownloader extend Downloader

File: src/Driver/VersionResolver.php (hunk 1 above)
Comment (Gurdarrag) [reply src/Driver/VersionResolver.php:14]: Yep true

File: src/Command/Input/OperatingSystemOption.php (hunk 2 above)
Comment (Gurdarrag) [reply src/Command/Input/OperatingSystemOption.php:85]: psalm is drunked on my environment

File: src/Driver/ChromeDriver/JsonVersionResolver.php (hunk 4 above)
Comment (Gurdarrag) [reply src/Driver/ChromeDriver/JsonVersionResolver.php:89]: i just refactorize in one class because there is not a lot of difference between the two

**Comment by Gurdarrag** [comment]
that's better now i think

File: src/Driver/ChromeDriver/VersionResolver.php [hunk 5]
```diff
@@ -74,7 +77,32 @@ public function latest(): Version
...
+        if ($browser->version()->major() < self::MAJOR_VERSION_ENDPOINT_BREAKPOINT) {
+            return $this->httpClient->request('GET', $this->getBrowserVersionEndpoint($browser))->getContent();
+        }
+
+        $response = $this->httpClient->request('GET', self::VERSION_ENDPOINT_JSON);
+        $versions = $response->toArray();
+        if (! array_key_exists($browser->version()->toString(), $versions['builds'])) {
+            throw new Unsupported(
+                sprintf('There is no build for version : %s', $browser->version()->toString())
+            );
```
Comment (nvdbeek) [inline src/Driver/ChromeDriver/VersionResolver.php:102]: **Suggestion:** use UnexpectedValueException instead of Unsupported.
The Unsupported is intended for browser-specific support.

**Comment by nvdbeek** [comment]
**Praise:** thanks for fixing this issue!

File: src/Driver/ChromeDriver/Downloader.php [hunk 6]
```diff
@@ -147,17 +153,35 @@ private function downloadArchive(Driver $driver): string
...
-            return self::BINARY_WINDOWS;
-        }
-
-        if ($operatingSystem->equals(OperatingSystem::MACOS())) {
-            return self::BINARY_MAC;
-        }
-
-        if ($operatingSystem->equals(OperatingSystem::LINUX())) {
-            return self::BINARY_LINUX;
+        switch ($this->isJsonVersion($driver)) {
```
Comment (nvdbeek) [inline src/Driver/ChromeDriver/Downloader.php:156]: **Todo:** you need to change the DownloaderTest & VersionResolverTest to support this changes with newer versions. Also good to add new case with the major breaking version.

File: src/Driver/ChromeDriver/VersionResolver.php (hunk 5 above)
Comment (Gurdarrag) [reply src/Driver/ChromeDriver/VersionResolver.php]: done

File: src/Driver/ChromeDriver/Downloader.php (hunk 6 above)
Comment (Gurdarrag) [reply src/Driver/ChromeDriver/Downloader.php:156]: yep i added some tests

**Review by KevinVanSonsbeek — APPROVED** [review]

**Review by nvdbeek — APPROVED** [review]

File: tests/Driver/ChromeDriver/DownloaderTest.php [hunk 7]
```diff
@@ -111,6 +158,20 @@ private function mockFsAndArchiveExtractorForSuccessfulDownload(OperatingSystem
...
+                $extractPath .= 'chromedriver-win32/';
+            }
+
+            if ($operatingSystem->equals(OperatingSystem::MACOS())) {
+                $extractPath .= 'chromedriver-mac-x64/';
+            }
+
+            if ($operatingSystem->equals(OperatingSystem::LINUX())) {
+                $extractPath .= 'chromedriver-linux64/';
+            }
```
Comment (dbrekelmans) [inline tests/Driver/ChromeDriver/DownloaderTest.php:172]: I feel like this shouldn't be necessary if we rename the dir after downloading, so the rest of the code can function as it did before.

**Review by dbrekelmans — CHANGES_REQUESTED** [review]

File: src/Driver/ChromeDriver/Downloader.php (hunk 6 above)
[the remark this answers is not in the document]
Comment (Gurdarrag) [reply src/Driver/ChromeDriver/Downloader.php:156]: ok for that

File: src/Driver/ChromeDriver/Downloader.php [hunk 8]
```diff
@@ -200,6 +227,29 @@ private function getFilePath(string $location, OperatingSystem $operatingSystem)
...
+        }
+
+        if ($operatingSystem->equals(OperatingSystem::LINUX())) {
+            return implode(
+                DIRECTORY_SEPARATOR,
+                [$location, self::BINARY_LINUX_JSON, $this->getFileName($operatingSystem)]
+            );
+        }
+
+        return $location . DIRECTORY_SEPARATOR . $this->getFileName($operatingSystem);
```
[the remark this answers is not in the document]
Comment (dbrekelmans) [reply src/Driver/ChromeDriver/Downloader.php:245]: Do you know why it fails? I don't see why it shouldn't work.

[the remark this answers is not in the document]
Comment (Gurdarrag) [reply src/Driver/ChromeDriver/Downloader.php:245]: Because of the / char i think

[the remark this answers is not in the document]
Comment (dbrekelmans) [reply src/Driver/ChromeDriver/Downloader.php:245]: Could you look a bit more into this? The full path that is being renamed already contains a `DIRECTORY_SEPARATOR`, so I don't see why adding more directory separators would suddenly break it. It would have a positive impact on the maintainability if we could solve this in the right place.

**Comment by oleg-andreyev** [comment]
any updates?

**Comment by Kamhal24** [comment]
Hello @dbrekelmans and @Gurdarrag. I should have some spare times in next week, is it ok if I try to fix those last 2 issues in PR? If so, should I just commit into those branch, or create MR into this branch?

**Comment by dbrekelmans** [comment]
Yes, by all means @Kamhal24!

**Comment by Kamhal24** [comment]
@Gurdarrag was faster :slightly_smiling_face:  fyi, not only google create new endpoints, but the archive now contains folder with the driver. Older versions had just the driver without folder. In case anyone needed this info someday

**Comment by Gurdarrag** [comment]
Sorry i was off from a month. 
I think it's better now, I just move driver file after extraction, to the root of archive folder
